Add netconf_get module

ganeshrn opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Implements part-1 of proposal 
https://github.com/ansible/proposals/issues/104
~Fixes https://github.com/ansible/ansible/issues/38819~
Fixes https://github.com/ansible/ansible/issues/40256

*  Add netconf_get module
*  Refactor `get`, `get_config`, `lock`, `unlock`
   and `discard_changes` netconf plugin api's
*  Add netconf module_utils file which netconf module
   related common functions
*  Refactor junos and iosxr netconf plugins
<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request


##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
netconf_get

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.6
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

ganeshrn committed:
Add netconf_get module

Implements part-1 of proposal #104
https://github.com/ansible/proposals/issues/104

*  Add netconf_get module
*  Refactor `get`, `get_config`, `lock`, `unlock`
   and `discard_changes` netconf plugin api's
*  Add netconf module_utils file which netconf module
   related common functions
*  Refactor junos and iosxr netconf plugins

ganeshrn committed:
Fix source option handling

ganeshrn committed:
Fix review comments

ganeshrn committed:
Update botmeta file

wisotzky commented:
As per my comments earlier, I would recommend the following changes in netconf_get.py.
[netconf_get.py.zip](https://github.com/ansible/ansible/files/1992684/netconf_get.py.zip)

ganeshrn commented:
@wisotzky Thanks for your patch. I tested it against `junos` and `iosxr` and it fails with error
```
    "changed": false,
    "invocation": {
        "module_args": {
            "filter": null,
            "format": "xml",
            "lock": null,
            "source": "running"
        }
    },
    "msg": "syntax error"
}
```
The problem seems because junos and iosxr does not support locking running datatstore. That's the reason I have added a check in my PR for `source=candidate`, if yes only then call lock and unlock. 
Also, I don't see much value add in locking the running datastore before fetching current active configuration. It is same as executing `#show running-config` in other <network-os>_command modules, it doesn't execute lock on running datastore before fetching acitve configuration.

wisotzky commented:
@ganeshrn, the reason why locking `running` is for consistency with some vendor implementations. For extra-large configurations, some NETCONF stacks need several minutes to render/transmit the complete configuration. If during this period another client is committing changes, it might happen that the XML rendered becomes inconsistent. Therefore the saver approach is to disallow others to change the configuration, while getting it.

The use-case for `candidate` is different. To be on the save side, you would lock the `candidate` before you do changes. So the assumption would be, that Ansible already has locked the `candidate` when doing running the `get-config` operation. Also there are other vendor implementations, like private candidates (Juniper) or implicit lock for candidate (Nokia) - where you don't need to lock the `candidate` datastore.

As a conclusion, it makes most sense to automatically lock `running` but not automatically lock `candidate`. My proposal allows you to explicitly override this default behavior, so you could decide NOT to lock running or to lock candidate.

If a vendor implementation does not support lock on `running`, it basically violates rfc6241. As per RFC, it does NOT even depend on the `:writable-running` feature. So I would not care if JUNOS or Cisco IOS-XR are creating errors if the "lock" option is not set. I've tested against Cisco/Tail-f ConfD and Nokia SROS - which both work nicely. Unfortunately I don't have currently access to Juniper/Cisco nodes from my setup.

ganeshrn commented:
>If a vendor implementation does not support lock on running, it basically violates rfc6241. As per RFC, it does NOT even depend on the `:writable-running` feature. So I would not care if JUNOS or Cisco IOS-XR are creating errors if the "lock" option is not set. I've tested against Cisco/Tail-f ConfD and Nokia SROS - which both work nicely. Unfortunately, I don't have currently access to Juniper/Cisco nodes from my setup.

Exactly it doesn't support `:writeable-running` feature but based on rfc content it has nothing to do with `<get-config>` operation
```

8.2.  Writable-Running Capability

8.2.1.  Description

   The :writable-running capability indicates that the device supports
   direct writes to the <running> configuration datastore.  In other
   words, the device supports <edit-config> and <copy-config> operations
   where the <running> configuration is the target.
```
So `lock` on running datstore while fetching `get-config` is not mandatory.

ganeshrn commented:
As a middle ground, we can keep `lock` option which defaults to false and add a capability in individual platform plugin say `supoort_lock_running` which will indicate if the platform supports lock on running datastore. So user if the user inputs `lock: True` and `source: running` based on platform capability it will throw an appropriate error for user to take corrective action ie. either set `lock` to `False` (default value) or change `source` to `candidate`.
Thoughts/Suggestions?

ganeshrn committed:
Update review comments and add support for lock

ganeshrn commented:
@wisotzky Based on above discussion and your patch I have updated the PR, please retest in your environment. To be in sync with other network get (command) modules the return value has `stdout` which is the raw XML received from the remote host. `xml` key will contain the transformed XML string with namespace removed from it so that users can apply plane `XPath` on this key value. The `output` will contain the formatted return values, currently supports `json` and `pretty`.

ganeshrn committed:
Lock update fix

ganeshrn committed:
Fix CI issue

wisotzky commented:
@ganeshrn, to return 3 different encodings sounds like quite some overhead to me. Imagine a configuration snapshot of a PE router could get >100MB. Unclear if the system is optimized to process and pass that amount of data - especially 3x more just because of those specifics.

I am also a bit sceptic about the namespaces. To apply the XPATH statement while ignoring the XMLNS should be a capability of the XML module. Also most XPATH processors allow you to add the namespace in curly brackets. And using the star wildcard, you could basically ignore the namespace. example /{*}state/{*}schema would search for /state/schema ignoring all namespaces.

ganeshrn committed:
Add integration test and minor fixes

ganeshrn commented:
> I've just said, that lock on running is the saver option to ensure consistency of data. 

Agreed that's why I mentioned in an earlier comment that `lock` option is added to the module with default `False`. Executing `lock` operation before `get_config` is not mandatory (even Netconf rfc doesn't mandate it). If the user wants to ensure consistency of data he can set `lock:True`, in that case the `source` datastore will be locked before fetching configuration. 
As per the updated PR if the `source` datastore doesn's support lock it will report an appropriate error.

Also, I have updated the PR with integration test run against `junos` and `iosxr` platforms.

ganeshrn committed:
Fix review comments

ganeshrn commented:
```
to return 3 different encodings sounds like quite some overhead to me. Imagine a configuration snapshot of a PE router could get >100MB. 
```
Based on your input I have modified the return as
```
    result = {
        'stdout': xml_resp,
        'output': output
    }
```
So now `stdout` will hold raw `xml` response including namespace and `output` will have formatted response with valid formats `json`, `pretty` and `xml`. For `display:xml` the output will have xml string with namespace removed. This result format is similar to `junos_command` (and other network get modules) and till no memory issues are reported with this return format.

> I am also a bit sceptic about the namespaces. To apply the XPATH statement while ignoring the XMLNS should be a capability of the XML module.
As `stdout` will have the raw response by default, the user will have the option to apply `xpath` with namespace included.

ganeshrn committed:
Fix CI failure

ganeshrn committed:
Fix CI issues

ganeshrn committed:
Fix CI issues

wisotzky commented:
Please have a deeper look into the code samples I've provided.
The following section comes to late in your code proposal
```
        if not HAS_JXMLEASE:
            module.fail_json(msg='jxmlease is required to display response in json format'
                                 'but does not appear to be installed. '
                                 'It can be installed using `pip install jxmlease`')
```
So actually you are sending the NETCONF get/get-config request to the server, before figuring out you cannot process the result because of jxmlease missing. This does not make much sense to me. Checks should be done, before you actually run the request.

wisotzky commented:
I am slowly getting confused, how selective you are applying my hardened module code into the your branch. You should not execute the locked_config, conn.get_config or conn_get without proper exception handling. Please remind, that ncclient always uses exceptions to inform the caller about what is going wrong.

Also remind, that the get operation also provides back a copy of the running config. So if the playbook author decides to take a lock, this should also apply to the get config operation. Therefore my code makes much more sense here.

wisotzky commented:
To fix all items listed above, please integrate the complete source files as attached.
It also includes the fix, to make Nokia SROS nodes working with connection=netconf.
[netconf_get.zip](https://github.com/ansible/ansible/files/2002459/netconf_get.zip)

Once you agree on those changes, I will help you with the authoring of the necessary netconf_rpc and netconf_lock modules.

wisotzky commented:
Please also revisit the netconf_config module! It might be meaningful to apply the same logic around "lock" for this module.

ganeshrn commented:
> Please note, that it is wrong to make lock on running dependent on :writable-running. If you lock the running datastore, you could even block any other copy/commit operation. 

I am not sure what you want to convey here. If the Netconf server doesn't support :writable-running capability than it cannot support `lock` operation on `running` datastore. If it doesn't support lock on running datastore there is no point in issuing `lock` on running datastore and catch the error message as execption.

> Please be aware, that lock on running is by far not an optional rfc6241 feature. So if a device does not support this, it must be considered as non-compliant to the NETCONF RFC.

That's wrong observation if the Netconf server doesn't support :writeable-running capability it won't support lock on running datatore, hence you can't claim it as non-complaint to Netconf RFC. It can support lock on candidate datastore based on  :candidate capability. This is the check I have added in `netconf/__init__.py` that is to figure out which all datastore's are supported based on server capability.

> Please also note, that typically controllers like ODL would typically lock both candidate and running datastore.

Again if Netconf server doesn't support direct write on `running` datastore it makes no sense to lock `running` datastore as it will result in error from Netconf server.

ganeshrn commented:
> So actually you are sending the NETCONF get/get-config request to the server, before figuring out you cannot process the result because of jxmlease missing.

Point taken, appreciate if you add it as a review comment in this PR.

ganeshrn commented:
> I am slowly getting confused, how selective you are applying my hardened module code into the your branch. 

Your hardened module is not working with the platforms that I am testing against, so I have to change it in such a way that it works across different Netconf implementation.

> You should not execute the locked_config, conn.get_config or conn_get without proper exception handling.

The `with locked_config` block ensures that `lock` and `unlock` are executed while entering and leaving the block respectively even if execption occurs. 

>  Please remind, that ncclient always uses exceptions to inform the caller about what is going wrong.

The execption raised by ncclient is already handled [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/network/common/netconf.py#L83)

I don't get why you want to catch `ConnectionError` in your module code unless you want to handle the execption in a different way.

FYI the `with locked_config` logic is based on junos modules that runs over Netconf transport using ncclient and has been stable since 2.3. This also helps with code reusability 


> Also remind, that the get operation also provides back a copy of the running config. So if the playbook author decides to take a lock, this should also apply to the get config operation. Therefore my code makes much more sense here.

Agreed, will add `get` call also with `with` block. 
FYI will continue to use `with` block due to reasons mentioned above.

wisotzky commented:
I am not sure, about your comment:

> That's wrong observation if the Netconf server doesn't support `:writeable-running` capability it won't support lock on running datatore, hence you can't claim it as non-complaint to Netconf RFC. It can support `lock` on candidate datastore based on :candidate capability. This is the check I have added in `netconf/__init__.py` that is to figure out which all datastore's are supported based on server capability.

What NETCONF RFC are you speaking about? According to RFC6241, the `running` datastore and the `lock` and `unlock` operation are mandatory to implement. As you can see in the part I've copied from this RFC, the target leafs for startup and candidate have dependencies: `if-feature candidate;` and `if-feature startup;` - while there is no dependency for running.

If you would be right, you would need to have a similar conditional statement `if-feature writable-running;` - which basically is not the case.

So in consequence lock/unlock support for the running datastore must be considered as mandatory functionality for a NETCONF server. This is even compliant to other statements in the same RFC.

The obvious question raised by you is about why should someone lock the running datastore, if the NETCONF server has not implemented the `:writable-running` feature or in other words, the running configuration is considered read-only. The answer is likely, that the writable-running capability is just about direct write access to the running datastore, using either the `edit-config` or `copy-config` operations with running as target datastore. Still, there might be other operations to modify the running datastore - for example doing a commit on candidate. According to the RFC description, doing a lock on running should also block the commit operation - independent if the NETCONF server supports the writable-running or not.

If a NETCONF server does not support lock/unlock on running, I would consider this as violation of rfc6241. But having this said, rfc6241 contains a lot of implementation details - and I have not yet seen/tested a single NETCONF server supporting all mandatory and optional features described in that IETF standard.

My conclusion would be, to keep it with the user and nodal implementation - if he wants to lock the datastore or not. If the user wants to lock running and the node supports it - why would the Ansible module block this with a corresponding error message on behalf of the node? Just to render a better error message? Not really convinced that this approach makes any sense...

ganeshrn committed:
Fix review comments and update integration test

wisotzky commented:
> Your hardened module code is not working with the platforms that I am testing against, so I have to change it in such a way that it works across different Netconf implementation.

Please indicate, what parts are not working... In my last delivery, the only different implementation is about lock - as I am not making any assumptions (correct or wrong) on existence of writable-running.

> The with `locked_config` block ensures that `lock` and `unlock` are executed while entering and leaving the block respectively even if execption occurs while executing within block.

I've seen this earlier in your branch, so I was aware about this functionality. The reason, why I did not added this into my patch, was mainly because I did not wanted to duplicate the part within the locked_config block - as this is somewhat the consequence of making the locking configurable. Also I believe there is a need to have an exception handler anyway - to raise appropriate error messages. So it seemed easier, to do explicit lock/unlock - instead of leveraging the shorter syntax of with locked_config...

Another note: the current locked_config code is always doing unlock, even if it was the lock in the first place raising an exception. Situation could be, that someone else holds the lock - so in NETCONF you would actually try to lock and after this you would do the unlock. I don't think this makes much sense - as you basically cause a second NETCONF error (for the unlock).

> The execption raised by ncclient is already handled here. I don't get why you want to catch ConnectionError in your module code unless you want to handle the execption in a different way.

I don't really see, how this exception is handled... The code reference you've added is about raising this Exception but not handling this Exception. In my testing, using your code I always ended up, that NETCONF errors are reported like this:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ansible.module_utils.connection.ConnectionError: None
fatal: [127.0.0.1]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible_PpI3Wd/ansible_module_netconf_get.py\", line 245, in <module>\n    main()\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible_PpI3Wd/ansible_module_netconf_get.py\", line 220, in main\n    response = conn.get(filter=filter_spec)\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible_PpI3Wd/ansible_modlib.zip/ansible/module_utils/network/common/netconf.py\", line 76, in __rpc__\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible_PpI3Wd/ansible_modlib.zip/ansible/module_utils/network/common/netconf.py\", line 108, in parse_rpc_error\nansible.module_utils.connection.ConnectionError: None\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
```

While I wanted a more simple error message like:
```
fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "Unexpected session close"}
```

I don't really like the error message to be "MODULE FAILURE" - as this is quite to generic. But indeed, it looks more that the error message parser actually causes the issue. So potentially might be better to fix this.

wisotzky commented:
Here another error message example from my setup.
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was:             Unlock operation is not supported with the current router configuration mode
fatal: [127.0.0.1]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible__r89FH/ansible_module_netconf_get.py\", line 245, in <module>\n    main()\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible__r89FH/ansible_module_netconf_get.py\", line 216, in main\n    with locked_config(module, target=source):\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/contextlib.py\", line 17, in __enter__\n    return self.gen.next()\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible__r89FH/ansible_modlib.zip/ansible/module_utils/network/netconf/netconf.py\", line 66, in locked_config\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible__r89FH/ansible_modlib.zip/ansible/module_utils/network/netconf/netconf.py\", line 57, in unlock_configuration\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible__r89FH/ansible_modlib.zip/ansible/module_utils/network/common/netconf.py\", line 76, in __rpc__\n  File \"/var/folders/x_/18gvwq452vl8gyj6y_t5qj780000gn/T/ansible__r89FH/ansible_modlib.zip/ansible/module_utils/network/common/netconf.py\", line 108, in parse_rpc_error\nansible.module_utils.connection.ConnectionError: \n            Unlock operation is not supported with the current router configuration mode\n        \n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
```

I am not sure, if it is desired not to catch the exception within the module, keeping it to Ansible to catch it. Especially to include all the traceback can be quite confusing - if this is a regular NETCONF RPC error message. For this specific case, wouldn't it just be enough to say "Unlock operation is not supported with the current router configuration mode". And this is what I've actually implemented.

Also, keep in mind that this trace was created using the `with locked_config` solution. I've debuted the NETCONF messages, and I can see two RPC error messages are generated. One for the failed lock and another for the failed unlock. So not only, that this code unnecessarily runs into the 2nd RPC error - also the failure message to the user indicated that the unlock was not successful - while actually the issue the user should know about is the failed lock.

ganeshrn commented:
>  the current locked_config code is always doing unlock, even if it was the lock in the first place raising an exception. Situation could be, that someone else holds the lock - so in NETCONF you would actually try to lock and after this you would do the unlock. I don't think this makes much sense - as you basically cause a second NETCONF error (for the unlock).

If user set `lock:True` in that case both lock and unlock needs to be executed so it make sense to have it in with block. If datastore is already locked (say by executing junos_rpc (TBD)) in that case the default value of `lock` is False and the with block won't be executed in first place, so scenario you  mentioned won't happen for netconf_get.

> The code reference you've added is about raising this Exception but not handling this Exception. In my testing, using your code I always ended up, that NETCONF errors are reported like this:

Can you please share how to reproduce the error message. In any case generic error handling should be done at a commonplace say (module_uitls.network.common.netconf) and not in each and every Netconf module.

wisotzky commented:
> If user set `lock:True` in that case both lock and unlock needs to be executed so it make sense to have it in with block. If datastore is already locked (say by executing junos_rpc (TBD)) in that case the default value of lock is False and the with block won't be executed in first place, so scenario you mentioned won't happen for netconf_get.

That's not the scenario I am after. Let's assume you have multiple management systems/controllers accessing the same NETCONF node. It might happen, that another NETCONF client currently owns the lock (temporary). In such case, you would either retry locking several times - or you would give up. But it does not make sense to unlock. See in my other post, we even generate the error message for unlock but not for lock.

Also note, that (in compliance to rfc6241) it could happen that another management protocol owns the lock. For example for nodes, which use the same candidate for CLI and NETCONF. So if a CLI operator currently is doing changes in the candidate, you would not be able to lock candidate from NETCONF.

ganeshrn committed:
Fix review comments

ganeshrn commented:
Alright, I have removed the with block and moved it to try..expect in `module_utils.network.netconf` so that `get` and `get_config` api's can be reused by other Netconf modules as well.

Please retest again, also feel free to add your integration test in `test/integration/targets/netconf_get/tests/sros` folder though we won't be able to test it in our infra as we don't have access to sros instance.

For details regarding running of integration test please refer
http://docs.ansible.com/ansible/devel/dev_guide/testing_integration.html#network-tests

wisotzky commented:
You did it almost right! Moving the functionality might be a very reasonable approach.
But don't really get, what you do [here](https://github.com/ganeshrn/ansible/blob/d305d43a70265b81ef5413f63c319d77a3e5df5e/lib/ansible/module_utils/network/netconf/netconf.py#L94)

Actually, the `get` method should not have the datastore `attribute:source` configurable. In case of the NETCONF get operation, this is alway accessing the running datastore. So this mapping should be hard-coded. Currently you are passing the value "None" - which likely will lead to errors.

Also I am seeing, that you have not used my updated documentation section. I've done some corrections, for example source can not be used with auto anymore, but is still described. Please just do a simple UNIX diff (or git diff) to see my changes. Also I've added examples for Nokia SROS.

Finally, I am struggling alway to get my updates for SROS back in... You've done a great job to update the NETCONF drivers for Cisco IOSXR and JUNOS. But to have it running with my Nokia nodes, I always need to add back my code changes. It absolutely make sense, to make this part of the same delivery.

Please let me know, if you are struggling to get Nokia nodes. From my knowledge, Nokia ION has provided the RedHat Ansible team with Nokia VSR software and license keys for development and regression testing (CI). Even this was mainly to build/support SROS CLI modules - clearly this could also be used for NETCONF regression. Let me clarify, what is the last VSR version shared with the Ansible team, as I do all my integration/testing using an early SROS16.0 load.

privateip commented on lib/ansible/modules/network/netconf/netconf_get.py:
```diff
@@ -0,0 +1,234 @@
+        source=dict(choices=['running', 'candidate', 'startup']),
+        filter=dict(),
+        display=dict(choices=['json', 'pretty', 'xml']),
+        lock=dict(default=False, type=bool)
```
Change this argument from bool to choices with the following choices `[never, always, if-supported]`

In the `never` case, do not attempt to lock the conf data store
In the `always` case, try to lock the conf data store and error if the lock operation failes
In the `if-supported` case, try to lock the conf data store and continue unlocked if the lock operation fails

The default value should be `never`

ganeshrn replied:
Fixed

wisotzky commented:
@ganeshrn 

Please review my changes put into PR #40198. It has been validated with ConfD and Nokia SROS devices. Please validate with Cisco IOS XR and Juniper, as I don't have those nodes available at the moment.

wisotzky commented:
@ganeshrn 

Could you please indicate, how NETCONF generic RPC framework is supposed to work? Seems like the current implementation of the Ansible wrapper around ncclient only has foreseen a single attribute to be passed to equal the name of the RPC operation. While I think there is the expectation, that someone could execute operations requiring additional attributes.

Just wanted to add the RPC module to my PR #40198. But as consequence of the above, I am struggling to get it running...

ganeshrn committed:
Fix review comments

wisotzky commented:
Hi @ganeshrn

I've included now the method `dispatch` for connection="netconf", which is more conform to ncclient which not really exposes the `rpc` method. Based on this new method, I've done a new Ansible module to call custom RPCs called `netconf_rpc` which is now part of #40198 .

I am using `dicttoxml` to provide more flexibility on how the request payload is defined. Hoping you like it. Still, there is one challenge with dicttoxml, as it does not supported unfolded style for lists. So I have put a PR on dicttoxml as well, to support the new "fold_list" feature.

Please be aware, that I did not yet spend cycles to define tests for `netconf_rpc`. As I only have SROS nodes and ConfD simulation, I would require your support to define tests for Cisco and Juniper nodes.

ganeshrn commented:
@wisotzky I have updated the PR with latest review comments.

If the module looks good, please approve this PR so that I can merge it to devel.
`netconf_get` module, `junos` and `iosxr` netconf plugin will be committed as part of this PR

Please raise a separate PR or modify the other existing PR that you raised to add sros only related changes and test cases as you are solely working on them.

If you have already started working on `netconf_rpc` please raise it in a different PR which only consist of ` netconf_rpc` related changes and test cases.

wisotzky commented:
@ganeshrn 

Can you give me some guidance on how to split my PR into multiple PR's?

The problem really is, that there are dependencies regarding the NETCONF connection layer in this PR. So you would not really be able to use such different PRs without having the initial PR being integrated. Unclear, how to expose those cross-dependencies.

Clearly you could first merge your PR into the devel - and I would add my additions as dedicated PR's. Makes live much easier...

BTW, is there a simple way to compare your PR against my PR proposal? If you want me to approve your PR, this would make my life much easier... As discussed, all my changes are done by reason and I have merged them with yours quite frequently. So my expectation on your PR is, that all code is now exactly the same - beside you should have taken out those things, you don't want to have merged in the same delivery (SROS related things, netconf_rpc).

ganeshrn commented:
@wisotzky 
Here is what I propose
1) If you and @privateip are ok with this PR I can get this merged to devel today.
2) After that, you raise a new PR which has `sros` related changes only. As you will apply your changes on top of devel it will have the `nectont_rpc` module and the underlying dependent changes already.

For `netconf_rpc` module
1) Please raise `netconf_rpc` in a new PR and the required changes in sros plugin. After it is reviewed and approved we can get it merged. I will work with you on this new PR that you raise to ensure the module you added works with junos and ios.

2) After the above PR is merged I will raise a PR to add test cases for `junos` and `ios` platforms

Also to make things simpler please close PR #40198 and  https://github.com/ansible/ansible/pull/39130

Does the above proposal work for you?

ganeshrn commented:
> The problem really is, that there are dependencies regarding the NETCONF connection layer in this PR. So you would not really be able to use such different PRs without having the initial PR being integrated. Unclear, how to expose those cross-dependencies.

I sense some confusion here. FYI devel is a working branch and a single feature can be broken into multiple PR's. The stable-X.X are the release branches from which the actual release of Ansible is done.

Here is the schedule for 2.6 release
https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_6.rst#release-schedule

FYI we are targetting `netconf_get` module in 2.6 release and other Netconf generic modules are pushed in 2.7 roadmap
https://github.com/ansible/community/wiki/Network:Core-roadmap-2.7

wisotzky commented:
> Does the above proposal work for you?

Okay, let's do that!

ganeshrn committed:
Fix review comments

Fix reveiw comments

ganeshrn commented:
> While only 7 out of those 16 files are directly related to the new netconf_get module, the others are more belonging to NETCONF common/plugin to make the new module work.

As long as there is a single module in PR and other supporting files it should be fine and in sync with module development guideline of Ansible. As you worked soley on `sros` plugin it makes sense to raise it in different PR.

@privateip Thank you for review, I will merge this PR now.

@wisotzky Thank you for working with me and testing the patch in your environment, as discussed earlier please raise a PR with `sros` plugin changes and we can get it merged immediately.

wrt to `netconf_rpc` we have time till 25th May 2018 to get it in 2.6
I will test your `netconf_rpc` PR in my environment with `junos` and `iosxr` and will try to get it in before module freeze date.

wisotzky commented:
@ganeshrn 
I've created a new PR #40358 which contains the new netconf_rpc module
